src/lib.rs

```diff
@@ -0,0 +1,531 @@
+mod jsonschema;
+
+use serde_json::{json, Map, Value};
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::iter::FromIterator;
+
+#[derive(Eq, PartialEq, Hash, Clone, Copy)]
+enum JSONSchemaKind {
+    Null,
+    Boolean,
+    Integer,
+    Number,
+    String,
+    Object,
+    Array,
+    AllOf,
+    OneOf,
+    Unknown,
+}
+
+struct JSONSchemaType {
+    kind: JSONSchemaKind,
+    nullable: bool,
+}
+
+impl JSONSchemaType {
+    pub fn new() -> Self {
+        JSONSchemaType {
+            kind: JSONSchemaKind::Unknown,
+            nullable: false,
+        }
+    }
+
+    /// Create typing context about a JSONSchema object
+    ///
+    /// # Arguments
+    /// * `value` - A serde_json::Value containing a schema
+    pub fn from_value(node: &Value) -> JSONSchemaType {
+        match &node["type"] {
+            Value::String(dtype) => {
+                let kind = JSONSchemaType::kind_from_string(dtype.as_str());
+                JSONSchemaType {
+                    kind: kind,
+                    nullable: kind == JSONSchemaKind::Null,
+                }
+            }
+            Value::Array(multitype) => {
+                let mut set: HashSet<JSONSchemaKind> = HashSet::from_iter(
+                    multitype
+                        .into_iter()
+                        .map(|s| JSONSchemaType::kind_from_string(s.as_str().unwrap().into())),
+                );
+                let mut json_type = JSONSchemaType::new();
+                if set.contains(&JSONSchemaKind::Null) {
+                    json_type.nullable = true;
+                    set.remove(&JSONSchemaKind::Null);
+                };
+                if set.len() > 1 {
+                    json_type.kind = JSONSchemaKind::Unknown;
+                } else {
+                    json_type.kind = set.into_iter().next().unwrap();
+                }
+                json_type
+            }
+            _ => {
+                let kind = if !node["properties"].is_null() {
+                    JSONSchemaKind::Object
+                } else if !node["allOf"].is_null() {
+                    JSONSchemaKind::AllOf
+                } else if !node["oneOf"].is_null() {
+                    JSONSchemaKind::OneOf
+                } else {
+                    JSONSchemaKind::Unknown
+                };
+                JSONSchemaType {
+                    kind: kind,
+                    nullable: false,
+                }
+            }
+        }
+    }
+
+    fn kind_from_string(dtype: &str) -> JSONSchemaKind {
+        match dtype {
+            "null" => JSONSchemaKind::Null,
+            "boolean" => JSONSchemaKind::Boolean,
+            "integer" => JSONSchemaKind::Integer,
+            "number" => JSONSchemaKind::Number,
+            "string" => JSONSchemaKind::String,
+            "object" => JSONSchemaKind::Object,
+            "array" => JSONSchemaKind::Array,
+            _ => panic!(),
+        }
+    }
+
+    pub fn as_bq_type(&self) -> String {
```

Comment:
Note: I removed this in a recent PR. But it could, my mental model of string allocation is clearer now. The bigquery module now uses serde for mapping strings to enums and back.

Reply:
👍